src/chain/fork_clone_builder.rs

```diff
@@ -30,65 +32,57 @@ pub trait ForkCloneBuilder<Response> {
+                            builder.commands.spawn(UnusedTarget).id()
+                        },
+                    )*
+                ];
 
-    fn build_fork_clone(
-        self,
-        source: Output<R>,
-        builder: &mut Builder,
-    ) -> Self::Outputs {
-        let target_0 = builder.commands.spawn(UnusedTarget).id();
-        let target_1 = builder.commands.spawn(UnusedTarget).id();
-
-        builder.commands.add(AddOperation::new(
-            Some(source.scope()),
-            source.id(),
-            ForkClone::<R>::new(
-                ForkTargetStorage::from_iter([target_0, target_1])
-            )
-        ));
-
-        let u_0 = (self.0)(Chain::new(target_0, builder));
-        let u_1 = (self.1)(Chain::new(target_1, builder));
-        (u_0, u_1)
+                builder.commands.add(AddOperation::new(
+                    Some(source.scope()),
+                    source.id(),
+                    ForkClone::<R>::new(
+                        ForkTargetStorage::from_iter(targets)
+                    )
+                ));
+                let ($($F,)*) = self;
+                // The compiler throws a warning when implementing this for
+                // tuple sizes that wouldn't use the result of the first _idx = _idx + 1
+                // so we add a leading underscore to suppress the warning
+                let mut _idx = 0;
+                (
+                    $(
+                        {
+                            let res = ($F)(Chain::new(targets[_idx], builder));
+                            _idx = _idx + 1;
+                            res
+                        },
+                    )*
+                )
+            }
+        }
     }
 }
 
-impl<R, F0, U0, F1, U1, F2, U2> ForkCloneBuilder<R> for (F0, F1, F2)
-where
-    R: 'static + Send + Sync + Clone,
-    F0: FnOnce(Chain<R>) -> U0,
-    F1: FnOnce(Chain<R>) -> U1,
-    F2: FnOnce(Chain<R>) -> U2,
-{
-    type Outputs = (U0, U1, U2);
-
-    fn build_fork_clone(
-        self,
-        source: Output<R>,
-        builder: &mut Builder,
-    ) -> Self::Outputs {
-        let target_0 = builder.commands.spawn(UnusedTarget).id();
-        let target_1 = builder.commands.spawn(UnusedTarget).id();
-        let target_2 = builder.commands.spawn(UnusedTarget).id();
-
-        builder.commands.add(AddOperation::new(
-            Some(source.scope()),
-            source.id(),
-            ForkClone::<R>::new(
-                ForkTargetStorage::from_iter([target_0, target_1, target_2])
-            )
-        ));
-
-        let u_0 = (self.0)(Chain::new(target_0, builder));
-        let u_1 = (self.1)(Chain::new(target_1, builder));
-        let u_2 = (self.2)(Chain::new(target_2, builder));
-        (u_0, u_1, u_2)
-    }
-}
+// Implements the `ForkCloneBUilder` trait for all tuples between size 2 and 15
+// (inclusive)
+all_tuples!(impl_forkclonebuilder_for_tuple, 2, 15, F, U);
```

Comment:
Very minor nitpick, but let's cap the tuple size to 12 for all of them for the sake of consistency, so users don't get additionally confused by hitting different tuple limits for different functions. If someone wants more than 12 forked clones there are several ways they could still accomplish that with a limit of 12.

Reply:
https://github.com/open-rmf/bevy_impulse/pull/2/commits/903663c7d31ef50da5b69198011cd001c3e3dad7 also found that there was a tuple with size 1 implemented manually and removed that to make the diff even more red https://github.com/open-rmf/bevy_impulse/pull/2/commits/c97a7fcd82c5a355c44514616a89399c916c65c3